source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -526,49 +526,51 @@ int ContextImpl::verifyCallback(X509_STORE_CTX* store_ctx, void* arg) {
   ContextImpl* impl = reinterpret_cast<ContextImpl*>(arg);
   SSL* ssl = reinterpret_cast<SSL*>(
       X509_STORE_CTX_get_ex_data(store_ctx, SSL_get_ex_data_X509_STORE_CTX_idx()));
-  Envoy::Ssl::SslExtendedSocketInfo* sslExtendedInfo =
+  return impl->doVerifyCertChain(
+      store_ctx,
       reinterpret_cast<Envoy::Ssl::SslExtendedSocketInfo*>(
-          SSL_get_ex_data(ssl, ContextImpl::sslExtendedSocketInfoIndex()));
+          SSL_get_ex_data(ssl, ContextImpl::sslExtendedSocketInfoIndex())),
+      bssl::UniquePtr<X509>(SSL_get_peer_certificate(ssl)),
+      static_cast<const Network::TransportSocketOptions*>(SSL_get_app_data(ssl)));
+}
 
-  if (impl->verify_trusted_ca_) {
+int ContextImpl::doVerifyCertChain(
+    X509_STORE_CTX* store_ctx, Ssl::SslExtendedSocketInfo* ssl_extended_info,
+    bssl::UniquePtr<X509> leaf_cert,
```

Comment:
I think this could take an `X509&`, possibly `const X509&`; I don't see a reason that this function should take ownership of leaf_cert.